internal/resource-request-operator/broadcaster.go

```diff
@@ -0,0 +1,183 @@
+
+import (
+	"fmt"
+	"sync"
+
+	crdclient "github.com/liqotech/liqo/pkg/crdClient"
+
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+
+	configv1alpha1 "github.com/liqotech/liqo/apis/config/v1alpha1"
+	"github.com/liqotech/liqo/pkg/utils"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+// Broadcaster is an object which is used to get resources of the cluster.
+type Broadcaster struct {
+	allocatable   corev1.ResourceList
+	clusterConfig configv1alpha1.ClusterConfig
+	offerMutex    sync.RWMutex
+	configMutex   sync.RWMutex
+}
+
+// SetupBroadcaster create the informer e run it to signal node changes updating Offers.
+func (b *Broadcaster) SetupBroadcaster(clientset *kubernetes.Clientset, config *rest.Config, stopper chan struct{}) error {
+	b.allocatable = corev1.ResourceList{}
+	factory := informers.NewSharedInformerFactory(clientset, 0)
+	informer := factory.Core().V1().Nodes().Informer()
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    b.onAdd,
+		UpdateFunc: b.onUpdate,
+		DeleteFunc: b.onDelete,
+	})
+
+	config.ContentConfig.GroupVersion = &configv1alpha1.GroupVersion
+	config.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
+	config.UserAgent = rest.DefaultKubernetesUserAgent()
+	CRDclient, err := crdclient.NewFromConfig(config)
+	if err != nil {
+		return err
+	}
+	go utils.WatchConfiguration(b.setConfig, CRDclient, "")
+	go informer.Run(stopper)
+	return nil
+}
+
+func (b *Broadcaster) setConfig(configuration *configv1alpha1.ClusterConfig) {
+	b.configMutex.Lock()
+	defer b.configMutex.Unlock()
+	b.clusterConfig = *configuration
+}
+
+// react to a Node Creation/First informer run.
+func (b *Broadcaster) onAdd(obj interface{}) {
+	node := obj.(*corev1.Node)
+	if node.Status.Phase == corev1.NodeRunning {
+		toAdd := &node.Status.Allocatable
+		currentResources := b.allocatable.DeepCopy()
+
+		for resourceName, quantity := range *toAdd {
+			value := currentResources[resourceName]
+			if !quantity.IsZero() {
+				value.Add(quantity)
+			}
+			currentResources[resourceName] = value
+		}
+
+		if err := b.writeResources(currentResources); err != nil {
+			klog.Errorf("OnAdd error: unable to write allocatable of Node %s: %s", node.Name, err)
+		}
+	}
+}
+
+// react to a Node Update.
+func (b *Broadcaster) onUpdate(oldObj, newObj interface{}) {
+	oldNode := oldObj.(*corev1.Node)
+	newNode := newObj.(*corev1.Node)
+	oldNodeResources := oldNode.Status.Allocatable
+	newNodeResources := newNode.Status.Allocatable
+	currentResources := b.allocatable.DeepCopy()
+	if newNode.Status.Phase == corev1.NodeRunning {
+		// node was already Running, update with possible new resources.
+		if oldNode.Status.Phase == corev1.NodeRunning {
+			for resourceName, quantity := range newNodeResources {
+				oldQuantity := oldNodeResources[resourceName]
```

Comment:
This function seems to be very complex, and with similar/repeated codes, can we split it into different functions?

Then I would suggest something like this to simplify the logic:
* if old is running subtract the resources
* if new is running add the resources

we don't care about the combination of the two, if the result is different we will update the stored value

Reply:
And what about the case when old is running and new is running too but allocatable resources change? I think we have to update subtracting or adding the difference between old and new but probably I am wrong.

Reply:
ok, you are right